sdk/core/Azure.Core/src/ResponseClassifier.cs

```diff
@@ -57,5 +130,199 @@ public virtual bool IsErrorResponse(HttpMessage message)
+        //    Response response,
+        //    string? message = null,
+        //    string? content = null,
+        //    string? errorCode = null,
+        //    IDictionary<string, string>? additionalInfo = null,
+        //    Exception? innerException = null)
+        //{
+        //    var formatMessage = CreateRequestFailedMessageWithContent(response, message, content, errorCode, additionalInfo);
+        //    var exception = new RequestFailedException(response.Status, formatMessage, errorCode, innerException);
+
+        //    if (additionalInfo != null)
+        //    {
+        //        foreach (KeyValuePair<string, string> keyValuePair in additionalInfo)
+        //        {
+        //            exception.Data.Add(keyValuePair.Key, keyValuePair.Value);
+        //        }
+        //    }
+
+        //    return exception;
+        //}
+
+        //public async ValueTask<string> CreateRequestFailedMessageAsync(Response response, string? message, string? errorCode, IDictionary<string, string>? additionalInfo, bool async)
+        //{
+        //    var content = await ReadContentAsync(response, async).ConfigureAwait(false);
+        //    return CreateRequestFailedMessageWithContent(response, message, content, errorCode, additionalInfo);
+        //}
+        #endregion
+
+        #region Exception details extraction & message creation methods
+        /// <summary>
+        /// </summary>
+        /// <param name="response"></param>
+        /// <param name="message"></param>
+        /// <param name="content"></param>
+        /// <param name="errorCode"></param>
+        /// <returns></returns>
+        internal string CreateRequestFailedMessageWithContent(Response response, string? message, string? content, string? errorCode)//, IDictionary<string, string>? additionalInfo)
+        {
+            StringBuilder messageBuilder = new StringBuilder()
+                .AppendLine(message ?? DefaultMessage)
+                .Append("Status: ")
+                .Append(response.Status.ToString(CultureInfo.InvariantCulture));
+
+            if (!string.IsNullOrEmpty(response.ReasonPhrase))
+            {
+                messageBuilder.Append(" (")
+                    .Append(response.ReasonPhrase)
+                    .AppendLine(")");
+            }
+            else
+            {
+                messageBuilder.AppendLine();
+            }
+
+            if (!string.IsNullOrWhiteSpace(errorCode))
+            {
+                messageBuilder.Append("ErrorCode: ")
+                    .Append(errorCode)
+                    .AppendLine();
+            }
+
+            // Note: we can ignore additionalInfo for now (for the purposes of the discussion of this design)
+            // because it's only passed in by exception creation methods called by the service libraries,
+            // and isn't useful when we are creating the exception from the response directly.
```

Comment:
This points to an opportunity to make some changes to the `ClientDiagnostics` API when we make it public in `ResponseClassifier`.  We can clarify what's strictly required (i.e. the response), and what available input properties are there to allow service libraries to customize the exception.